Use mach-nix to add zephyr requirements to nix-shell

dxmh opened the pull request:
As per your suggestion in https://github.com/zmkfirmware/zmk/pull/1026#issuecomment-985883000, use [`mach-nix`](https://github.com/DavHau/mach-nix/blob/master/examples.md#mkpython--mkpythonshell) to handle the python environment, reading zephyr's python dependencies from `zephyr/scripts/requirements.txt`.

dxmh committed:
fix(setup): Add zephyr requirements to nix-shell

Use mach-nix to handle zephyr's python requirements.

bromanko commented:
Thanks for doing this! I think we will need to revise it. The `zephyr` dir gets added by West. So, I think we need to either run the west commands automatically when the shell is initialized, or leave those steps and the pip install command as manual. Thoughts?

dxmh commented:
Ah, thanks for pointing that out - I hadn't spotted the `zephyr` directory gets added by west.

I notice in the ZMK container the `requirements.txt` file is [read from upstream](https://github.com/zmkfirmware/zmk-docker/blob/25972e7aad57a2b48f6141185711943cf76fc03b/Dockerfile#L26-L27):

```shell
pip3 install  -r https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/v${ZEPHYR_VERSION}/scripts/requirements-base.txt
```

We could consider doing something similar for this mach-nix approach? 

We'd need to ensure the correct version of zephyr is referenced. In the ZMK docker container this is stored [in a variable in the job that builds the container image](https://github.com/zmkfirmware/zmk-docker/blob/25972e7aad57a2b48f6141185711943cf76fc03b/.github/workflows/containers.yml#L4) (which gets [version bumped when necessary](https://github.com/zmkfirmware/zmk-docker/commit/af6c5091aad887f0c46ea686579dd9a6bf2ec317#diff-bdcc6a2a85f645f62724fe8dafbf0581cb0c1d65f6a76cb2985a9172e31a473c)):

```yaml
env:
  zephyr-version: 2.5.0
  ...
```

As the shell.nix would be in the main ZMK repo, maybe we could get the version from somewhere in the local repo already? For example:

https://github.com/bromanko/zmk/blob/a8c7cf4f19ad80eb8e6148f872d3a9584b25befc/app/west.yml#L12

What do you think about this approach of fetching the remote requirements.txt file? Might need some further consideration...

dxmh committed:
fix(setup): Fetch Zephyr requirements from URL

The requirements cannot be read from the file in `./zephyr/` because
this directory does not yet exist in a clean repo.

Fetching the requirements file from an upstream URL is consistent with
how the ZMK docker image does it:

  https://github.com/zmkfirmware/zmk-docker/blob/25972e7aad57a2b48f6141185711943cf76fc03b/Dockerfile#L26-L27

dxmh committed:
refactor(setup): More descriptive variable name

dxmh commented:
I made some adjustments to fetch the `requirements-base.txt` file from a remote URL. 

I believe there is likely a better approach this, especially with some of the discussion going on in https://github.com/zmkfirmware/zmk/pull/1026. However, I thought it would be worth updating this PR in case it prompts any further discussion, and because with these changes to your original PR, I was able to get up and running with ZMK on a new NixOS machine. 

Firmware built successfully with the following commands:

```console
[laptop:~/code/zmk]$ nix-shell
[nix-shell:~/code/zmk]$ west init -l app/
[nix-shell:~/code/zmk]$ west update
[nix-shell:~/code/zmk]$ west zephyr-export
[nix-shell:~/code/zmk/app]$ west build --pristine --board nice_nano -- -DSHIELD=a_dux_left -DZMK_CONFIG=/home/dom/code/zmk-config/config
```
